File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        # Do a heap from the lists O(nlog(n))
        heap = []
        for linkedList in lists:
            while(linkedList is not None):
                heappush(heap, linkedList.val)
                linkedList = linkedList.next

        if not heap:
            return None
            
        # Then create a linked list O(n)
        head = ListNode(heappop(heap))
        node = head
        while(heap):
            nextElt = heappop(heap)
            node.next = ListNode(nextElt)
            node = node.next
        return head

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
         PriorityQueue<ListNode> pq = new PriorityQueue<>((a,b) -> a.val-b.val);
        for(ListNode list:lists)
        {
            if(list!=null)
            {
                pq.add(list);
            }
            
        }
        ListNode head=null;
        ListNode tail=null;
        
        while(!pq.isEmpty())
        {
            ListNode current=pq.poll();
            if(head==null)
            {
                head=current;
                tail=current;
                
            }
            else
            {
                tail.next=current;
                tail=current;
                
            }
            if(current.next!=null)
                pq.add(current.next);
        }
        return head;
}
}